### DISCUSSION TITLE
v6 Type 'string' is not assignable to type 'never' when select fields

### DISCUSSION THREAD

**Discussion by imcm7** [mikro-orm/mikro-orm#4742] [General]
> Type 'string' is not assignable to type 'never'.

```
    69 |       { active: true },
    70 |       {
  > 71 |         fields: ['id', 'name', 'roles.name'],
       |                  ^^^^
    72 |         populate: ['roles'],
    73 |       }
    74 |     );
```

```
async findAndCountAllPermissions <P extends string = never> (
    query: FilterQuery<Permission>,
    options?: FindOptions<Permission, P>
  ): Promise<{  count: number, items: Array<Permission> }> {

              * * *

  }
```

**Comment by B4nan** [comment]
Just like the `P` arg type, you need to have one for the `fields` hint:

```ts
async findAndCountAllPermissions <P extends string = never, F extends string = never> (
    query: FilterQuery<Permission>,
    options?: FindOptions<Permission, P, F>
  ): Promise<{  count: number, items: Array<Permission> }> {
```

**Reply by imcm7** [reply]
Why so much stuff

**Reply by B4nan** [reply]
That's the only way to make the partial loading type safe, you need type args to be able to pass the information from arguments to the return type.

**Reply by imcm7** [reply]
@B4nan 

```
 let permission: Loaded<Permission, P, F> | null = null;
 permission = await repository.findOne(query, options); // work without options with type Permission not Loaded<Permission, P, F>
 repository.assign(permission, data); // here Type 'Selected<Permission, P, F>' is missing the following properties from type 'Permission': name, slug, isInitialized, isTouched, and 10 more.ts(2345)
 ```
 
 I can search and filter by relations without populate the entity?

**Reply by B4nan** [reply]
Hmm, that's just a type-level issue, will take a look if I can quickly fix that in v6. For now, you can do this:

```ts
repository.assign(permission as Permission, data);
```

**Reply by B4nan** [reply]
This should do it:

https://github.com/mikro-orm/mikro-orm/commit/21ebc0fd2c6730b7511eccda14b2dfcb468f4904

edit: this was much more complicated, but I managed to find a way: https://github.com/mikro-orm/mikro-orm/commit/053cb46a304ac8736262baac7a7fc91d6a33d3be?w=1

**Reply by imcm7** [reply]
And now

```
Type '{ count: number; items: Selected<Permission, "roles", "id" | "name" | "roles.name">[]; }' is not assignable to type '{ count: number; items: Permission[]; }'.
  Types of property 'items' are incompatible.
```

I import `import { Selected } from '@mikro-orm/core/typings';` but what to do with it?

Selected<Permission, '', 'id'>

May be https://github.com/mikro-orm/mikro-orm/commit/21ebc0fd2c6730b7511eccda14b2dfcb468f4904  will help with that too

**Reply by B4nan** [reply]
Partially loaded results are indeed not assignable to the full entity type, that's the whole point of this feature.

**Reply by imcm7** [reply]
I not understand how to use partially loaded result, I not understand what to put in Selected<Permission, ??, ??> in controller, like that P, F in service.

**Reply by imcm7** [reply]
I understand how to use it `Selected<Permission, "roles", "id" | "name" | "roles.name">`.
Thanks, now all looks good

**Comment by imcm7** [comment]
Troubles after update to `dev 6.0.0-dev.108`. 
I resolved 27 problems but remains 2. One is:

> Property 'find' does not exist on type 'Selected<Collection<Role, object>, never, "name">'


```
permissions = await permissionService.findAndCountAllPermissions(
      { active: true },
      {
        fields: ['id', 'name', 'roles.name'],
        populate: ['roles'],
      }
); // that return { count: 0, items: [] }

for (const permission of permissions['items']) {
   permission.roles!.find()
}
```

and 

> Type '"roles.permissions"' is not assignable to type '"email" | "password" |

```
const user: Selected<User,  'roles.permissions'> | null = await this.userService.findUser(  // service return type Promise<Loaded<User, P, F> | null>

// I try
// Loaded<User,  'roles.permissions'> 
// and Loaded<User, unknown,  'roles.permissions'>  and other.

      { email, active: true },
      {
        populate: [
          'roles.permissions',
        ],
      }
```

**Reply by B4nan** [reply]
How does the `findAndCountAllPermissions` method look like? I just did a quick test on master and working with collections seems to be fine. Your `roles` property should be typed to something like `LoadedCollection<Selected<Role, never, "name">>`. Do you have the code live so I can check it?

Btw I just rebased all the fixes from v5, so better to update to `6.0.0-dev.110`

**Reply by imcm7** [reply]
```
async findAndCountAllPermissions <P extends string = never, F extends string = never> (
    query: FilterQuery<Permission>,
    options?: FindOptions<Permission, P, F>
  ): Promise<{ 
    count: number,
    items: Array<Loaded<Permission, P, F>>
  }> {
    let result: [Array<Loaded<Permission, P, F>>, number] =  [[], 0];

    try {
      result = await this.permissionRepository.findAndCount(query, options);
    } catch (error: any) {
      throw new Error('Unable to load permissions and count.');
    }

    return {
      count: result[1],
      items: result[0],
    };
  }

```

**Reply by imcm7** [reply]
I get result of `await roleService.findAllRoles()` like `roles: Selected<Role, never, "id" | "name">[]`

```
  async findAllRoles<P extends string = never, F extends string = never> (
    query: FilterQuery<Role>,
    options?: FindOptions<Role, P, F>
  ): Promise<Array<Loaded<Role, P, F>>> {
```

**Reply by B4nan** [reply]
You need to use `dev.110`, not `dev.109`, that version wasn't published completely due to NPM outage.

**Reply by B4nan** [reply]
It works fine for me, just tried in the realworld example repo.

**Reply by imcm7** [reply]
I reinstall second tim and is ok. Don't know what that was, because I installed manually all pckages and set to all the same version form terminal 6.0.0-dev.110

**Reply by imcm7** [reply]
now all is ok
